src/overlay/OverlayManagerImpl.cpp

```diff
@@ -532,42 +532,49 @@ OverlayManagerImpl::tick()
+        // Exit early: no pending slots available
+        return;
+    }
+
     auto availableAuthenticatedSlots = availableOutboundAuthenticatedSlots();
-    auto nonPreferredCount = nonPreferredAuthenticatedCount();
+    bool allOutboundArePreferred = nonPreferredAuthenticatedCount() == 0;
 
-    // if all of our outbound peers are preferred, we don't need to try any
-    // longer
-    if (availableAuthenticatedSlots > 0 || nonPreferredCount > 0)
+    // First, attempt to replace all connections with preferred peers
+    if (!allOutboundArePreferred)
     {
-        // try to replace all connections with preferred peers
-        auto pendingUsedByPreferred = connectTo(
-            mApp.getConfig().TARGET_PEER_CONNECTIONS, PeerType::PREFERRED);
+        auto preferredToConnect = std::min(
+            availablePendingSlots,
+            static_cast<int>(mApp.getConfig().TARGET_PEER_CONNECTIONS));
+        auto pendingUsedByPreferred =
+            connectTo(preferredToConnect, PeerType::PREFERRED);
 
         assert(pendingUsedByPreferred <= availablePendingSlots);
         availablePendingSlots -= pendingUsedByPreferred;
+    }
 
-        // connect to non-preferred candidates from the database when
-        // PREFERRED_PEER_ONLY is set and we connect to a non preferred_peer we
-        // just end up dropping & backing off it during handshake (this allows
-        // for preferred_peers to work for both ip based and key based preferred
-        // mode)
-        if (availablePendingSlots > 0 && availableAuthenticatedSlots > 0)
-        {
-            // try to leave at least some pending slots for peer promotion
-            constexpr const auto RESERVED_FOR_PROMOTION = 1;
-            auto outboundToConnect =
-                availablePendingSlots > RESERVED_FOR_PROMOTION
-                    ? std::min(availablePendingSlots - RESERVED_FOR_PROMOTION,
-                               availableAuthenticatedSlots)
-                    : RESERVED_FOR_PROMOTION;
-            auto pendingUsedByOutbound =
-                connectTo(outboundToConnect, PeerType::OUTBOUND);
-            assert(pendingUsedByOutbound <= availablePendingSlots);
-            availablePendingSlots -= pendingUsedByOutbound;
-        }
+    // Second, if there is capacity for pending and authenticated outbound
+    // connections, connect to more peers. Note: connect even if
+    // PREFERRED_PEER_ONLY is set, to support key-based preferred peers mode
+    // (see PREFERRED_PEER_KEYS). When PREFERRED_PEER_ONLY is set and we connect
+    // to a non-preferred peer, drop it and backoff during handshake.
+    if (availablePendingSlots > 0 && availableAuthenticatedSlots > 0)
+    {
+        // try to leave at least some pending slots for peer promotion
+        constexpr const auto RESERVED_FOR_PROMOTION = 1;
+        auto outboundToConnect =
+            availablePendingSlots > RESERVED_FOR_PROMOTION
+                ? std::min(availablePendingSlots - RESERVED_FOR_PROMOTION,
```

Comment:
@MonsieurNicolas I didn't change this calculation, as I'm still not sure about the contract between outbound vs inbound-to-outbound promotions. Based on the discussion in #1753 (comment), it seems like we should prioritize outbound target before attempting promotions, but this is not what the code is doing here (as it subtracts the const from pending before calculating min)